components/_impl/workers/subprocess_worker.py

```diff
@@ -0,0 +1,346 @@
+                assert result == subprocess_rpc.BOOTSTRAP_INPUT_LOOP_SUCCESS, result
+
+                self._worker_bootstrap_finished = True
+                assert self._proc.poll() is None
+            except (Exception, KeyboardInterrupt) as e:
+                stdout, stderr = get_output()
+                cause = "import failed" if self._proc.poll() else "timeout"
+                raise e from RuntimeError(
+                    f"Failed to bootstrap worker ({cause}):\n"
+                    f"    working_dir: {self.working_dir}\n"
+                    f"    stdout:\n{textwrap.indent(stdout, ' ' * 8)}\n\n"
+                    f"    stderr:\n{textwrap.indent(stderr, ' ' * 8)}"
+                )
+
+    def _log_cmd(self, snippet: str) -> None:
+        with open(self._command_log, "at", encoding="utf-8") as f:
+            now = datetime.datetime.now().strftime("[%Y-%m-%d] %H:%M:%S.%f")
+            f.write(f"# {now}\n{snippet}\n\n")
+
+    @contextlib.contextmanager
+    def watch_stdout_stderr(self):
+        # Get initial state for stdout and stderr, since we only want to
+        # capture output since the contextmanager started.
+        stdout_stat = os.stat(self._stdout_f.name)
+        stderr_stat = os.stat(self._stderr_f.name)
+
+        def get() -> typing.Tuple[str, str]:
+            with open(self._stdout_f.name, "rb") as f:
+                _ = f.seek(stdout_stat.st_size)
+                stdout = f.read().decode("utf-8").strip()
+
+            with open(self._stderr_f.name, "rb") as f:
+                _ = f.seek(stderr_stat.st_size)
+                stderr = f.read().decode("utf-8").strip()
+
+            return stdout, stderr
+
+        yield get
+
+    def _run(self, snippet: str) -> None:
+        """Helper method for running code in a subprocess."""
+        assert self._worker_bootstrap_finished
+        assert self.alive, "Process has exited"
+
+        snippet = textwrap.dedent(snippet)
+        with self.watch_stdout_stderr() as get_output:
+            self._input_pipe.write(snippet.encode(subprocess_rpc.ENCODING))
+            self._log_cmd(snippet)
+
+            result = marshal.loads(self._output_pipe.read())
+            if isinstance(result, str):
+                assert result == subprocess_rpc.SUCCESS
+                return
+
+            assert isinstance(result, dict)
+            serialized_e = subprocess_rpc.SerializedException(**result)
+            stdout, stderr = get_output()
+            subprocess_rpc.SerializedException.raise_from(
+                serialized_e=serialized_e,
+                extra_context=(
+                    f"    working_dir: {self.working_dir}\n"
+                    f"    stdout:\n{textwrap.indent(stdout, ' ' * 8)}\n\n"
+                    f"    stderr:\n{textwrap.indent(stderr, ' ' * 8)}"
+                )
+            )
+
+    def _kill_proc(self) -> None:
+        """Best effort to kill subprocess."""
+        if getattr(self, "_proc", None) is None:
```

Comment:
Can we just set self._proc = None at the start of the ctor? Seems more intuitive.

Reply:
Ah, but then it would be an `Optional[subprocess.Popen]` and we'd have to none check everywhere when it goes back to Pytorch proper and has to pass the type checker...

I would consider this getattr pattern at least as Pythonic as setting a default at the start of the ctor.